typed/merge.go

```diff
@@ -280,72 +291,71 @@ func (w *mergingWalker) doList(t *schema.List) (errs ValidationErrors) {
 	return errs
 }
 
-func (w *mergingWalker) visitMapItems(t *schema.Map, lhs, rhs *value.Map) (errs ValidationErrors) {
-	out := &value.Map{}
+func (w *mergingWalker) visitMapItem(t *schema.Map, out map[string]interface{}, key string, lhs, rhs value.Value) (errs ValidationErrors) {
+	fieldType := t.ElementType
+	if sf, ok := t.FindField(key); ok {
+		fieldType = sf.Type
+	}
+	pe := fieldpath.PathElement{FieldName: &key}
+	w2 := w.prepareDescent(pe, fieldType)
+	w2.lhs = lhs
+	w2.rhs = rhs
+	errs = append(errs, w2.merge(pe.String)...)
+	if w2.out != nil {
+		out[key] = *w2.out
+	}
+	w.finishDescent(w2)
+	return errs
+}
+
+func (w *mergingWalker) visitMapItems(t *schema.Map, lhs, rhs value.Map) (errs ValidationErrors) {
+	out := map[string]interface{}{}
 
 	if lhs != nil {
-		for i := range lhs.Items {
-			litem := &lhs.Items[i]
-			fieldType := t.ElementType
-			if sf, ok := t.FindField(litem.Name); ok {
-				fieldType = sf.Type
-			}
-			w2 := w.prepareDescent(fieldpath.PathElement{FieldName: &litem.Name}, fieldType)
-			w2.lhs = &litem.Value
+		lhs.Iterate(func(key string, val value.Value) bool {
+			var rval value.Value
 			if rhs != nil {
-				if ritem, ok := rhs.Get(litem.Name); ok {
-					w2.rhs = &ritem.Value
+				if item, ok := rhs.Get(key); ok {
+					rval = item
 				}
 			}
-			if newErrs := w2.merge(); len(newErrs) > 0 {
-				errs = append(errs, newErrs...)
-			} else if w2.out != nil {
-				out.Items = append(out.Items, value.Field{litem.Name, *w2.out})
+			errs = append(errs, w.visitMapItem(t, out, key, val, rval)...)
+			if rval != nil {
+				rval.Recycle()
```

Comment:
optional nit: can `defer` this when rval is set?

Reply:
done